### DISCUSSION TITLE
Window size

### DISCUSSION THREAD

**Discussion by ddkwork** [cogentcore/core#795] [Q&A]
I seem to see a way to set the window size in the goosi driver window object, but the goosi window interface doesn't seem to provide the signature. In general, every app should have a way to set the size, but at the moment I haven't found this method, if there is a way to set the size of the main window, please let me know, if not, whether it is allowed to add a window interface signature to operate it.

**Comment by kkoreilly** [comment]
I will look into this soon.

**Comment by kkoreilly** [comment]
You can specify a minimum default size for the window by setting the `Min` style property of the body. For example:

```go
func main() {
	b := gi.NewAppBody("My App")
	b.Style(func(s *styles.Style) {
		s.Min.Set(units.Dp(500), units.Dp(300))
	})
        b.NewWindow().Run().Wait()
}
```

**Reply by kkoreilly** [reply]
Also, I made it so that windows are at least half of the screen size on desktop by default.

**Comment by ddkwork** [comment]
> You can specify a minimum default size for the window by setting the Min style property of the body. For example:\n\nfunc main() {\n\tb := gi.NewAppBody(\"My App\")\n\tb.Style(func(s *styles.Style) {\n\t\ts.Min.Set(units.Dp(500), units.Dp(300))\n\t})\n        b.NewWindow().Run().Wait()\n}

Thanks you.

**Comment by ddkwork** [comment]
> You can specify a minimum default size for the window by setting the `Min` style property of the body. For example:
> 
> ```go
> func main() {
> 	b := gi.NewAppBody("My App")
> 	b.Style(func(s *styles.Style) {
> 		s.Min.Set(units.Dp(500), units.Dp(300))
> 	})
>         b.NewWindow().Run().Wait()
> }
> ```

```go
func main() {
	b := gi.NewAppBody("My App")
	b.Style(func(s *styles.Style) {
		s.Min.Set(units.Dp(500), units.Dp(300))
	})
	w := b.NewWindow()
	w.SetFullWindow(true)
	w.Run().Wait()
}
```

The setting size and full screen still do not take effect, the first run takes effect, the subsequent run is still a small window, and w.SetFullWindow(true) cannot be full screen

**Reply by kkoreilly** [reply]
`w.SetFullWindow(true)` does not set a window as fullscreen; instead, it sets it as taking up all of the space of the parent window, which has no effect when it is a new system window. I will work on adding support for setting windows as fullscreen soon in the context of #27.

**Reply by kkoreilly** [reply]
By setting the minimum size of the body, you are specifying a default minimum value for the size of the window. However, after the user resizes the window, it saves that as their preferred size and will use that by default until they resize it again.

**Reply by ddkwork** [reply]
> By setting the minimum size of the body, you are specifying a default minimum value for the size of the window. However, after the user resizes the window, it saves that as their preferred size and will use that by default until they resize it again.

I will try again, thx you

**Comment by ddkwork** [comment]
> You can specify a minimum default size for the window by setting the `Min` style property of the body. For example:
> 
> ```go
> func main() {
> 	b := gi.NewAppBody("My App")
> 	b.Style(func(s *styles.Style) {
> 		s.Min.Set(units.Dp(500), units.Dp(300))
> 	})
>         b.NewWindow().Run().Wait()
> }
> ```

![Screenshot_20240118_010401_com.realvnc.viewer.android.jpg](https://github.com/cogentcore/core/assets/19886504/c53dc9bf-8a72-4810-826e-098b4ed367d5)

Maybe I don't understand what you mean, setting the minimum value of the body style to resize the main window has always succeeded once, and each subsequent one has failed. As for the full-screen scenario, there are these: packet grabbers, debugger programs, etc., which require large windows or even DND mode.

**Reply by kkoreilly** [reply]
For some reason I can not reproduce this issue. Firstly, if you are using the latest version of Cogent Core (`go get cogentcore.org/core@main`), windows should automatically take up at least half the width and height of the window, so you should not need to set the min size of the window anymore. Also, as long as you do not resize the window, any min size you do set should take effect. Once you resize the window, it will save that resized size as your preference and make it the default size of that window in the future.

**Reply by ddkwork** [reply]
> For some reason I can not reproduce this issue. Firstly, if you are using the latest version of Cogent Core (go get cogentcore.org/core@main), windows should automatically take up at least half the width and height of the window, so you should not need to set the min size of the window anymore. Also, as long as you do not resize the window, any min size you do set should take effect. Once you resize the window, it will save that resized size as your preference and make it the default size of that window in the future.

I tested it again this morning, and as you mentioned, when we press the left mouse button to widen the window, theoretically as long as we click the button to close the window normally, the size of the window we dragged should be saved to the app preferences, which is actually written into the appdata configuration directory. However, I didn't succeed in testing, either in terms of styling the appbody with code or dragging and dropping windows to save to the config file. Leave this question alone for now, and then we'll look back at it later, so let's test something else first

**Reply by kkoreilly** [reply]
I implemented automatic remembering of window fullscreen state, so if you make a window fullscreen, it will stay fullscreen when you run the app again.

**Reply by ddkwork** [reply]
![screenshots](https://github.com/cogentcore/core/assets/19886504/4230a1a5-fd34-482a-b76c-7e3f91e39ef9)
```go
GOROOT=C:\Users\Admin\go\pkg\mod\golang.org\toolchain@v0.0.1-go1.22.0.windows-amd64 #gosetup
GOPATH=C:\Users\Admin\go #gosetup
C:\Users\Admin\go\pkg\mod\golang.org\toolchain@v0.0.1-go1.22.0.windows-amd64\bin\go.exe build -o C:\Users\Admin\AppData\Local\JetBrains\GoLand2023.3\tmp\GoLand\___go_build_github_com_ddkwork_crypt.exe github.com/ddkwork/crypt #gosetup
C:\Users\Admin\AppData\Local\JetBrains\GoLand2023.3\tmp\GoLand\___go_build_github_com_ddkwork_crypt.exe
   Trace [2024-02-09 19:12:57]    --------- title --------- | ------------------ info ------------------ //runtime.doInit1 proc.go:7176
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x7ff79d2b6863]

goroutine 20 [running]:
cogentcore.org/core/goosi/driver/desktop.(*Window).Close(0x0)
        D:/workspace/workspace/core/goosi/driver/desktop/window.go:289 +0x23
created by cogentcore.org/core/goosi/driver/base.(*AppMulti[...]).QuitClean in goroutine 52
        D:/workspace/workspace/core/goosi/driver/base/app_multi.go:125 +0xb6

进程 已完成，退出代码为 2
```

**Reply by kkoreilly** [reply]
What did you do that led to this panic? Is it reliably reproducible?

**Comment by ddkwork** [comment]
It will be triggered as soon as the window is closed, and it will be triggered every time.



---Original---
From: ***@***.***&gt;
Date: Sat, Feb 10, 2024 14:38 PM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;"State ***@***.***&gt;;
Subject: Re: [cogentcore/core] Window size (Discussion #795)




 
What did you do that led to this panic? Is it reliably reproducible?

**Reply by kkoreilly** [reply]
How are you closing the window? It does not panic for me when I close it using the `Close window` button in the taskbar context menu. Does it still crash for you when you are not using the debugger? Does it still crash for you when closing the demo, or only when closing this specific app?

**Comment by ddkwork** [comment]
Whether it's the close button in the top right corner or the taskbar right-click close, it's triggered by any application. Why do you think the security check I suggest will never happen?
&nbsp;https://github.com/uber-go/nilaway
https://github.com/cogentcore/core/discussions/858



---Original---
From: ***@***.***&gt;
Date: Sat, Feb 10, 2024 14:48 PM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;"State ***@***.***&gt;;
Subject: Re: [cogentcore/core] Window size (Discussion #795)




 
How are you closing the window? It does not panic for me when I close it using the Close window button in the taskbar context menu. Does it still crash for you when you are not using the debugger? Does it still crash for you when closing the demo, or only when closing this specific app?

**Reply by kkoreilly** [reply]
I will look more into the possibility of using nilaway. Can you please answer whether this happens when running an app normally outside of GoLand? Also, can't you try running nilaway on the codebase and informing me of any errors that are produced?

**Reply by ddkwork** [reply]
![Uploading screenshots.gif…]()

**Comment by ddkwork** [comment]
It's the same when running outside of goland, for nilaway, I reported to you a null pointer scan in the discussion forum a long time ago, and I'm using it to check my golibrary repositories, as well as other repositories

https://github.com/cogentcore/core/discussions/745


---Original---
From: ***@***.***&gt;
Date: Sat, Feb 10, 2024 14:57 PM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;"State ***@***.***&gt;;
Subject: Re: [cogentcore/core] Window size (Discussion #795)




 
I will look more into the possibility of using nilaway. Can you please answer whether this happens when running an app normally outside of GoLand? Also, can't you try running nilaway on the codebase and informing me of any errors that are produced?

**Comment by ddkwork** [comment]
Run the hello example code of code directly to reproduce it



---Original---
From: ***@***.***&gt;
Date: Sat, Feb 10, 2024 14:38 PM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;"State ***@***.***&gt;;
Subject: Re: [cogentcore/core] Window size (Discussion #795)




 
What did you do that led to this panic? Is it reliably reproducible?

**Comment by ddkwork** [comment]
> I will look more into the possibility of using nilaway. Can you please answer whether this happens when running an app normally outside of GoLand? Also, can't you try running nilaway on the codebase and informing me of any errors that are produced?

![screenshots](https://github.com/cogentcore/core/assets/19886504/8cdfa87b-4245-4319-b8c4-91b8fef10373)

**Comment by ddkwork** [comment]
[nilaway.txt](https://github.com/cogentcore/core/files/14228855/nilaway.txt)

**Comment by kkoreilly** [comment]
I fixed the window closing crash in 13d8db3b72c6bbb6c67f651f7180edeb0c7388af.

**Comment by ddkwork** [comment]
> I fixed the window closing crash in 13d8db3b72c6bbb6c67f651f7180edeb0c7388af.

The full screen finally worked, and the setting size still didn't work

**Comment by ddkwork** [comment]
![big zoom](https://github.com/cogentcore/core/assets/19886504/0014c38c-5535-4309-9a8b-181e5104ddbc)

**Reply by kkoreilly** [reply]
`&goosi.Screen{ScreenNumber:0, Geometry:image.Rectangle{Min:image.Point{X:0, Y:0}, Max:image.Point{X:2560, Y:1600}}, DevicePixelRatio:2, PixSize:image.Point{X:5120, Y:3200}, PhysicalSize:image.Point{X:270, Y:170}, LogicalDPI:480, PhysicalDPI:481.65927, Depth:24, RefreshRate:60, Orientation:0, NativeOrientation:0, PrimaryOrientation:0, Name:"EDO PnP Monitor", Manufacturer:"", Model:"", SerialNumber:""}`

**Reply by ddkwork** [reply]
> `&goosi.Screen{ScreenNumber:0, Geometry:image.Rectangle{Min:image.Point{X:0, Y:0}, Max:image.Point{X:2560, Y:1600}}, DevicePixelRatio:2, PixSize:image.Point{X:5120, Y:3200}, PhysicalSize:image.Point{X:270, Y:170}, LogicalDPI:480, PhysicalDPI:481.65927, Depth:24, RefreshRate:60, Orientation:0, NativeOrientation:0, PrimaryOrientation:0, Name:"EDO PnP Monitor", Manufacturer:"", Model:"", SerialNumber:""}`

When I saw your last fix, it felt like the values weren't passed through the parameters as much as possible, which led me to another question: could it be that the recipients of a certain struct method are not uniform? Because I've seen a couple of places where there are methods that use pointer receiving, and other methods that do away with pointer and use struct receiving, then in this case it can only handle values, and when its parameters come from the parent method through a non-pointer operation, the expected value may not be passed.Goland will also give a hint that the struct's method recipients should use pointers for all methods, and clicking on its suggestion pop-up menu will add an * to all method recipients in bulk.

**Reply by kkoreilly** [reply]
All of the places that we use value receivers, it should be safe and necessary to use value receivers. We do not need to have everything be consistent in terms of whether there are pointer or value receivers for a certain type. If there are any specific cases in which you think a value receiver is causing problems, you can report that.

Also, I am working on the math for your screen scaling, and it would help me to know whether you have a zoom factor in your Windows display settings; can you send me a screenshot of your Windows display settings?

**Reply by ddkwork** [reply]
> All of the places that we use value receivers, it should be safe and necessary to use value receivers. We do not need to have everything be consistent in terms of whether there are pointer or value receivers for a certain type. If there are any specific cases in which you think a value receiver is causing problems, you can report that.
> 
> Also, I am working on the math for your screen scaling, and it would help me to know whether you have a zoom factor in your Windows display settings; can you send me a screenshot of your Windows display settings?

![{B2460C61-9E1A-4d8d-AD31-CDED8634E1EE}.png](https://github.com/cogentcore/core/assets/19886504/0d58e322-c022-4684-b04c-eaac80df6361)

**Reply by kkoreilly** [reply]
Thanks! I will look into this further.

**Reply by kkoreilly** [reply]
I did the math, and it seems like the app is scaled correctly on your device; it is just big because you have a scale factor of 200% set in your Windows settings. You can either decrease the scale factor in your Windows settings or in your Cogent Core settings, as you have done. If you have any other window sizing or scaling issues, please let me know.

**Comment by ddkwork** [comment]
I didn't pay attention to the zoom value under the Windows system settings before,When you asked me to check, I also saw that it was 200%,Generally speaking,100% should be the default value,I don't know if it's automatically set by the Win11 system or Huawei PC Manager is set to such a value when installing the driver。 So my initial idea was the same as you described, and it was normal to change 200% to 100%, but when I adjusted it to 100%, the whole window was quite small, and all the widgets on it were only the size of bees, which was too small. Another reason is as I described on AnyDesk before: I have also tested on Android phones, compared to tablets, the default display parameters of various systems of Android phones are stable, and I have not changed them, but it is also very large on Android phones, I have compiled 6 apks to test, and each apk has to be adjusted once zoom, otherwise it will not work. If I want, I can record a video for the troubles on Android to demonstrate that all 6 apks need to do this. At this point, my suggestion is still described in Anydesk: It's too troublesome to investigate this confusion, please give me a way to set the zoom and space of the current app in the main function, and I will manually set it in the main function of each app.



---Original---
From: ***@***.***&gt;
Date: Mon, Feb 12, 2024 08:53 AM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;"State ***@***.***&gt;;
Subject: Re: [cogentcore/core] Window size (Discussion #795)




 
I did the math, and it seems like the app is scaled correctly on your device; it is just big because you have a scale factor of 200% set in your Windows settings. You can either decrease the scale factor in your Windows settings or in your Cogent Core settings, as you have done. If you have any other window sizing or scaling issues, please let me know.

**Reply by kkoreilly** [reply]
Can you send me a screenshot of the demo on an Android device with default 100% zoom, and then a screenshot of some other standard app with similar widgets on the same device? It does not make any sense to manually set the settings in the main function; if there is something wrong with our sizing, we need to fix it.

**Comment by ddkwork** [comment]
In addition, setting the window size still doesn't work, yesterday on anydesk we didn't change the width and height value to test, after updating your submission, no matter what the size of the set width and height value is, this window is large and has not changed



---Original---
From: ***@***.***&gt;
Date: Mon, Feb 12, 2024 08:53 AM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;"State ***@***.***&gt;;
Subject: Re: [cogentcore/core] Window size (Discussion #795)




 
I did the math, and it seems like the app is scaled correctly on your device; it is just big because you have a scale factor of 200% set in your Windows settings. You can either decrease the scale factor in your Windows settings or in your Cogent Core settings, as you have done. If you have any other window sizing or scaling issues, please let me know.

**Reply by kkoreilly** [reply]
As I said previously, once the user runs the app, it will remember the window size and the window size all future times they run the app will be the same as they last set it. As long as the window size is staying the same as it was the last time the app was run, everything is working correctly. The min you set in the body styles specifies the default minimum size of the window before the user runs the app, but afterward it will remember the user's preferences.

**Comment by ddkwork** [comment]
I just updated your submission today.,It's normal on Windows.,Yesterday no matter how you dragged the window to resize.,It's big to run again.,Today I can actually save the size of the dragged or full screen.,Windows is finally normal.,Thank you.。 Wait for me a minute I'll set up the SDK and compile the apk test, haven't refreshed the apk for a long time



---Original---
From: ***@***.***&gt;
Date: Mon, Feb 12, 2024 10:25 AM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;"State ***@***.***&gt;;
Subject: Re: [cogentcore/core] Window size (Discussion #795)




 
Can you send me a screenshot of the demo on an Android device with default 100% zoom, and then a screenshot of some other standard app with similar widgets on the same device? It does not make any sense to manually set the settings in the main function; if there is something wrong with our sizing, we need to fix it.

**Comment by ddkwork** [comment]
Tested the latest submission and the apk is back to normal, thank you



---Original---
From: ***@***.***&gt;
Date: Mon, Feb 12, 2024 10:25 AM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;"State ***@***.***&gt;;
Subject: Re: [cogentcore/core] Window size (Discussion #795)




 
Can you send me a screenshot of the demo on an Android device with default 100% zoom, and then a screenshot of some other standard app with similar widgets on the same device? It does not make any sense to manually set the settings in the main function; if there is something wrong with our sizing, we need to fix it.
